### PR TITLE
[full nodes] Add support for multiple simultaneous network configs

### PR DESCRIPTION
**Pull request by bothra90**
## Motivation

With this PR, we will now instantiate as many network stacks on a single Libra node as are present in its configuration.

Additionally, 
- We split keypairs into network keypairs (stored in
NetworkConfig) and consensus keypair (stored in ConsensusConfig). This is needed because a node can have several network keypairs (one per network it is a part of), whereas it should have at most 1 consensus keypair.
- TrustedPeersConfig was conflating 3 separate concepts, which are now separated into different config options:
  1. Validators (HashMap<PeerId, Ed25519PublicKey>): Nodes in consensus
  2. NetworkTrustedPeers (HashSet<Ed25519PublicKey>): Nodes allowed to
   connect
  3. UpstreamPeers (HashSet<Ed25519PublickKey>): Nodes which are
   "upstream" in state synchronization.

Further cleanup:
- ConfigBuilder in LibraSwarm should allow creation of different types
  of a swarms: a validator swarm (currently allowed), a trusted full
  node swarm, and a public full node swarm.
- Remove notion of PeerId from network stack, since it's not possible to
  prove that a node has the PeerId it's claiming to have w/o a trusted
  peers map from network identity key to PeerId, which we do not have in
  permission-less networks.
- Generate a unique node id for a LibraNode that is different from
  PeerId, since a node can have more than 1 PeerId.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

We test to make sure there's no regression in running validator clusters. For now, full nodes are not being tested to any extent.

Local LibraSwarm: OK
Started 4 node cluster on AWS using terraform

## Related PRs
#736

### CODE REVIEW

File: config/src/config.rs [hunk 1]
```diff
@@ -110,169 +103,6 @@ impl Default for BaseConfig {
...
-    // getters for keypairs
-    pub fn get_network_identity_keypair(&self) -> (X25519StaticPrivateKey, X25519StaticPublicKey) {
-        (
-            self.get_network_identity_private(),
-            self.get_network_identity_public().clone(),
-        )
-    }
-}
-
 #[derive(Clone, Debug, Eq, PartialEq)]
```
Comment (phlip9) [inline config/src/config.rs]: derive Copy

File: config/src/config.rs [hunk 2]
```diff
@@ -589,6 +463,23 @@ impl Default for ConsensusConfig {
             contiguous_rounds: 2,
             max_pruned_blocks_in_mem: None,
             pacemaker_initial_timeout_ms: None,
+            consensus_keypair: ConsensusKeyPair::default(),
+            consensus_keypair_file: PathBuf::from("consensus_keypair.config.toml"),
+        }
+    }
+}
+
+#[cfg(any(test, feature = "testing"))]
```
Comment (phlip9) [inline config/src/config.rs]: from https://github.com/libra/libra/blob/master/documentation/coding_guidelines.md#testing I think you can replace this manual `Clone` impl with

```rust
#[cfg_attr(any(test, feature = "testing"), derive(Clone))]
#[derive(Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct ConsensusConfig {
    // ..
}
```

File: config/src/config.rs [hunk 3]
```diff
@@ -711,38 +610,24 @@ impl NodeConfig {
...
-                get_local_ip().ok_or_else(|| ::failure::err_msg("No local IP"))?;
+        let mut validator_count = 0;
+        for network in &mut config.networks {
+            // We use provided peer id for validator role. Otherwise peer id is generated using
+            // network identity key.
+            if network.role == "validator" {
+                assert_eq!(
+                    validator_count, 0,
+                    "At most 1 network config should be for a validator"
+                );
```
Comment (phlip9) [inline config/src/config.rs]: `assert!(peer_id.is_some());` ?

File: config/src/keys.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            network_identity_public_key,
+        }
+    }
+
+    /// Beware, this destroys the private key from this NodeConfig
+    pub fn take_network_signing_private(&mut self) -> Option<Ed25519PrivateKey> {
+        self.network_signing_private_key.take()
+    }
+
+    pub fn get_network_identity_private(&self) -> X25519StaticPrivateKey {
```
Comment (phlip9) [inline config/src/keys.rs]: does this need a testing feature gate, since it clones a private key?

File: config/src/keys.rs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        Self {
+            consensus_private_key: PrivateKeyContainer::Present(consensus_private_sig),
+            consensus_public_key: Some(consensus_public_sig),
+        }
+    }
+}
+
+impl ConsensusKeyPair {
+    // used to deserialize keypairs from a configuration file
+    pub fn load_config<P: AsRef<Path>>(path: P) -> Self {
```
Comment (phlip9) [inline config/src/keys.rs]: will we eventually return `Result<Self>` in these functions?

**Comment by sunmilee** [comment]
lgtm as well, so the plan is to add back the full node configs once we figure out separation of trusted peer configs to three different entities?

File: libra_node/src/main_node.rs [hunk 6]
```diff
@@ -200,12 +200,12 @@ pub fn setup_network(
...
             .map(|(peer_id, addrs)| (peer_id.try_into().expect("Invalid PeerId"), addrs))
             .collect();
-        let network_signing_private = config.peer_keypairs.take_network_signing_private()
+        let network_signing_private = config.network_keypairs.take_network_signing_private()
             .expect("Failed to move network signing private key out of NodeConfig, key not set or moved already");
         let network_signing_public: Ed25519PublicKey = (&network_signing_private).into();
         network_builder
             .transport(TransportType::TcpNoise(Some(
-                config.peer_keypairs.get_network_identity_keypair(),
+                config.network_keypairs.get_network_identity_keypair(),
```
Comment (phlip9) [inline libra_node/src/main_node.rs]: IIRC we shouldn't clone private keys. Don't need to handle it in this PR though.

/cc @kchalkias @huitseeker do we have any guidelines about in-memory cryptographic material? There's lots of other things we could be doing like guard pages or [clear-on-drop](https://github.com/cesarb/clear_on_drop).

File: network/src/connectivity_manager/mod.rs [hunk 7]
```diff
@@ -324,13 +324,18 @@ where
     fn handle_request(&mut self, req: ConnectivityRequest) {
         match req {
             ConnectivityRequest::UpdateAddresses(peer_id, addrs) => {
+                debug!(
+                    "Received updated addresses for peer: {}",
+                    peer_id.short_str()
+                );
```
Comment (phlip9) [inline network/src/connectivity_manager/mod.rs]: I think other network modules are using `trace!` for this kind of logging?

File: testsuite/tests/libratest/smoke_test.rs [hunk 8]
```diff
@@ -286,7 +280,7 @@ fn test_basic_state_synchronization() {
     assert!(swarm.add_node(node_to_restart.clone(), false).is_ok());
 
     // Wait for all the nodes to catch up
-    swarm.wait_for_all_nodes_to_catchup();
+    assert_eq!(true, swarm.wait_for_all_nodes_to_catchup());
```
Comment (phlip9) [inline testsuite/tests/libratest/smoke_test.rs]: replace `assert_eq!(true, _)` with `assert!(_)`

File: config/config_builder/src/swarm_config.rs [hunk 9]
```diff
@@ -5,52 +5,17 @@
...
 use failure::prelude::*;
 use std::path::{Path, PathBuf};
 
-/// Topology indicates the shape of the validator network
-/// Currently does not handle full nodes, the launch_swarm will only use num_nodes value
-#[derive(Debug, Clone)]
-pub struct LibraSwarmTopology {
-    // TODO: make it more general to support various network shapes
-    data: Vec<usize>,
-}
```
[the remark this answers is not in the document]
Comment (bothra90) [reply config/config_builder/src/swarm_config.rs]: we would, but this was not the right abstraction for it. Will bring this (or something similar) back once we know how we want to create swarms during tests.

File: config/src/keys.rs (hunk 4 above)
Comment (bothra90) [reply config/src/keys.rs]: yes

File: config/src/keys.rs (hunk 5 above)
Comment (bothra90) [reply config/src/keys.rs]: yeah that would be ideal. For now, I'm unifying all `load_config` and `save_config` functions into a single trait which will make it easier.

Comment (bothra90) [reply config/src/keys.rs]: Will take that up later though

File: common/executable_helpers/src/helpers.rs [hunk 10]
```diff
@@ -84,9 +78,9 @@ pub fn setup_executable(
         is_logging_disabled,
         Some(config.base.node_async_log_chan_size),
     );
-
-    setup_metrics(&config.network.peer_id, &config);
-
+    for network in &config.networks {
+        setup_metrics(&network.peer_id, &config);
```
[the remark this answers is not in the document]
Comment (bothra90) [reply common/executable_helpers/src/helpers.rs]: Since the peer id of a node will be different on different networkconfigs, I don't think we have the bucketing problem you're mentioning. From the code, I don't there's any issue in calling `setup_metrics` multiple times with different peer ids.

File: libra_node/src/main_node.rs (hunk 6 above)
Comment (bothra90) [reply libra_node/src/main_node.rs]: Will take care of this at a later point.

File: config/src/keys.rs [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+impl Default for NetworkKeyPairs {
+    fn default() -> Self {
+        let mut rng = StdRng::from_seed(TEST_SEED);
+        let (net_private_sig, net_public_sig) = compat::generate_keypair(&mut rng);
+        let (private_kex, public_kex) = x25519::compat::generate_keypair(&mut rng);
+        Self {
+            network_signing_private_key: PrivateKeyContainer::Present(net_private_sig),
+            network_signing_public_key: net_public_sig,
+            network_identity_private_key: private_kex,
+            network_identity_public_key: public_kex,
```
Comment (phlip9) [inline config/src/keys.rs]: public/private_kex => public/private_key

Comment (bothra90) [reply config/src/keys.rs]: name was added by @huitseeker, so I was going with the assumption that it was intentional to indicate that these are of type `x25519`.

**Comment by phlip9** [comment]
LGTM, though we should wait to land until we can confirm your cluster works

**Comment by bothra90** [comment]
Confirmed that cluster on AWS works and is able to accept and execute transactions.

File: terraform/validator-sets/dev/seed_peers.config.toml [hunk 12]
```diff
@@ -1,2 +1,2 @@
 [seed_peers]
-8deeeaed65f0cd7484a9e4e5ac51fbac548f2f71299a05e000156031ca78fb9f = ["/ip4/SEED_IP/tcp/6180"]
+1e5d5a74b0fd09f601ac0fca2fe7d213704e02e51943d18cf25a546b8416e9e1 = ["/ip6/::1/tcp/46267"]
```
Comment (mgorven) [inline terraform/validator-sets/dev/seed_peers.config.toml]: Revert this change
